Initial implementation of Tooltip Component

vipentti opened the pull request:
Initial implementation of a Tooltip component to resolve #54.

Example usage: 

```elm
            
import Material.Tooltip as Tooltip
import Material.Icon as Icon

-- Note the index in both the Render as well as the mouse event handlers

tooltip : Model -> Html Msg
tooltip model =
    div []
    [ Icon.view "add" [ Tooltip.attach Mdl [0] ]
    , Tooltip.render Mdl [0] model.mdl
        []
        [text "Default tooltip"]
    ]
                   
```

Currently there is only one problem that I'm not sure how to best solve: Currently tooltips are always contained in a `div` element which might not be always what you want. However, I'm not sure on the best way to allow users to select the container element for the tooltip.

One option would be  adding a configurable property to Config e.g.

```elm 
{-| Tooltip config
-}
type alias Config a =
  { size : Size
  , position : Position
  , container : HtmlElementFunction a 
  }


-- Set the element
container : HtmlElementFunction a -> Property m
container elem = 
  (\options -> { options | container = elem })



-- Usage 
tooltip : Model -> Html Msg
tooltip model =
    div []
    [ Icon.view "add" [ Tooltip.attach Mdl [0] ]
    , Tooltip.render Mdl [0] model.mdl
        -- Set the container to span element
        [ Tooltip.container Html.span ]
        [text "Default tooltip"]
    ]

```

Open to feedback and suggestions.

debois commented on src/Material/Tooltip.elm:
```diff
@@ -0,0 +1,523 @@
+    , right
+    , top
+    , bottom
+    , default
```
I don't think we need a `default` property: you specify it by not doing anything else. None of the other components have it.

debois commented on src/Material/Tooltip.elm:
```diff
@@ -0,0 +1,523 @@
+    , default
+    , large
+    , Property
+    , onMouseEnter
```
I don't think we should expose `onMouseEnter` / `onMouseLeave`; user's should always just use `attach`, yes?

vipentti replied:
`onMouseEnter` and `onMouseLeave` are needed if you want to have a tooltip for a regular `Html m` element.

debois commented on src/Material/Tooltip.elm:
```diff
@@ -0,0 +1,523 @@
+        Left ->
+          { left = props.left - offsetWidth - 10
+          , top =
+              if (top + marginTop < 0) then
```
Very difficult to read. Maybe write the `if` on a single line:

    , top = if (top + marginTop < 0) then 0 else top

Same multiple times below.

debois commented on src/Material/Tooltip.elm:
```diff
@@ -0,0 +1,523 @@
+        defaultPos
+  in
+    Options.styled div
+      (cs "mdl-tooltip"
```
No need to manually construct the list with `::`; I think the following is clearer:

    [ cs "mdl-tooltip" 
    , cs "is-active" `when` model.isActive
    , ...

debois commented on src/Material/Tooltip.elm:
```diff
@@ -0,0 +1,523 @@
+-}
+onMouseEnter : (Parts.Msg (Container b) -> d) -> Parts.Index -> Attribute d
+onMouseEnter lift idx =
+  Html.Events.on "mouseenter" (Json.map (Enter >> ((pack idx) >> lift)) stateDecoder)
```
This is such a clever use of the parts mechanics. I really like it.

debois commented on src/Material/Tooltip.elm:
```diff
@@ -0,0 +1,523 @@
+
+Refer to [this site](http://debois.github.io/elm-mdl#/tooltips)
+for a live demo.
+
```
Add docs here indicating that users must _both_ use `attach` to install handlers on the target element, as well as call `render` on the following sibling element.

debois commented on src/Material/Tooltip.elm:
```diff
@@ -0,0 +1,523 @@
+        ([ "target" ] ++ parents ++ [ "nextSibling" ])
+
+    paths =
+      List.map createPath [0..4]
```
You look for the tooltip element among the next 5 siblings, right? Is there are reason we don't just demand that it's the following element?

vipentti replied:
I think the problem is that if you have for example the following HTML using MDL tooltips

```html 
<p>HTML is related to but different from 
    <span id="xml"><i>XML</i></span>.
    <span>Here is another span</span>  
</p>
<span for="xml" class="mdl-tooltip">Tooltip</span>
```

Then in this case the first nextSibling for the target would be the other span, so we have to travel back the tree to find the actual tooltip element because there is no easy way to tell how nested the tooltips target actually is.

debois commented on src/Material/Tooltip.elm:
```diff
@@ -0,0 +1,523 @@
+{-| Decodes a DOMState from a DOM event
+-}
+stateDecoder : Json.Decoder DOMState
+stateDecoder =
```
Neat.

debois commented on src/Material/Tooltip.elm:
```diff
@@ -0,0 +1,523 @@
+  Options.set (\options -> { options | size = Large })
+
+
+{-| Maps a Html.Attribute to Material Property
```
Do we need this? I couldn't find a use of it, and I really don't want to give users the ability to install arbitrary attributes on elm-mdl constructs—I've seen them in the past use it in weird ways that break various things, then file bug reports here.

vipentti replied:
You are right, I apparently forgot to remove it.

debois commented on demo/Demo/Tooltip.elm:
```diff
@@ -0,0 +1,222 @@
+                      div []
+                        [ Icon.view "add" [ Tooltip.attach Mdl [0] ]
+                        , Tooltip.render Mdl [0] model.mdl
+                            [Tooltip.default]
```
Should just

     []

yes?

debois commented on demo/Demo/Tooltip.elm:
```diff
@@ -0,0 +1,222 @@
+    [Grid.size Grid.All 4]
+    [ div [style [("text-align", "center")]]
+        [tooltip]
+    , code description
```
Description should be 

    , p [] [ text description ] 

(Code-blocks should always be, well, code :)

debois commented on demo/Demo/Tooltip.elm:
```diff
@@ -0,0 +1,222 @@
+                   [Tooltip.default]
+                   [text "Default tooltip"]
+               ]
+            , """
```
Generally need more space between the element with the tooltip and the description below. Especially Button.

debois commented:
Very cool! 

I made a couple of comments on the source which you may or may not want to take into account. Take a look and let me know what you think?

vipentti commented:
Thanks a lot for the feedback! 

Implemented most of the suggestions and added support for changing the container element in config 

e.g. 

```elm

tooltip = 
  div []
     [ Icon.view "face"
         [ Tooltip.attach Mdl [4] ]     
     , Tooltip.render Mdl [4] model.mdl
       [ Tooltip.large
       , Tooltip.top
       , Tooltip.container Html.span]
         [ img [src "assets/elm.png", width 24, height 24] []
         , text "A target with a tooltip containing both an image and text."
         ]
     ]
```
